arbor/morph/morphology.cpp

```diff
@@ -0,0 +1,197 @@
+#include <cmath>
+#include <iostream>
+#include <unordered_map>
+#include <utility>
+
+#include <arbor/math.hpp>
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/morphology.hpp>
+#include <arbor/morph/sample_tree.hpp>
+#include <arbor/morph/primitives.hpp>
+
+#include "algorithms.hpp"
+#include "io/sepval.hpp"
+#include "util/span.hpp"
+#include "util/strprintf.hpp"
+
+namespace arb {
+
+namespace impl{
+
+std::vector<mbranch> branches_from_parent_index(const std::vector<msize_t>& parents,
+                                                const std::vector<point_prop>& props,
+                                                bool spherical_root)
+{
+    using util::make_span;
+
+    const char* errstr_single_sample_root =
+        "A morphology with only one sample must have a spherical root";
+    const char* errstr_incomplete_cable =
+        "A branch must contain at least two samples";
+
+    if (parents.empty()) return {};
+
+    auto nsamp = parents.size();
+
+    // Enforce that a morphology with one sample has a spherical root.
+    if (!spherical_root && nsamp==1u) {
+        throw morphology_error(errstr_single_sample_root);
+    }
+
+    std::vector<int> bids(nsamp);
+    int nbranches = spherical_root? 1: 0;
+    for (auto i: make_span(1, nsamp)) {
+        auto p = parents[i];
+        bool first = is_root(props[p]) || is_fork(props[p]);
+        bids[i] = first? nbranches++: bids[p];
+    }
+
+    std::vector<mbranch> branches(nbranches);
+    for (auto i: make_span(nsamp)) {
+        auto p = parents[i]==mnpos? 0: parents[i];
+        auto& branch = branches[bids[i]];
+
+        // Determine the id of the parent branch if this is the first sample in
+        // the branch, and include the fork/root point at the end of the parent
+        // branch where applicable.
+        // This is icky, but of all the solutions, this is the cleanest (that I
+        // could find).
+        if (!branch.size()) {
+            // A branch has null root if either:
+            //      ∃ a spherical root and branch id is 0
+            //      ∄ a spherical root and parent id is 0
+            auto null_root = spherical_root? !i: !p;
+            branch.parent_id = null_root? mnpos: bids[parents[i]];
+
+            // Add the first sample to a branch that is attached to
+            // non-spherical root if the branch is not branch 0.
+            if ((null_root && i) || is_fork(props[p])) {
+                branch.index.push_back(p);
+            }
+        }
```

Comment:
I think the above logic can a bit simplified. If we just leave `p` equal to `parents[i]` then
1. Our parent branch is `p==mnpos? mnpos: bids[p]`.
2. The only situations where `p` is not included in a new branch are: `p` is `mnpos`, or `p` is 0 and the root is spherical.

e.g.
```
branch.parent_id = p==mnpos? mnpos: bids[p];
if (p!=mnpos && !(p==0 && spherical_root)) {
    branch.index.push_back(p);
}
```

Reply:
Not so simple:
1. parent branch is also `npos` if  `!spherical_root && p==0`.

I have gone with the following, but I don't know if it saves much.
```
        auto p = parents[i];
        auto& branch = branches[bids[i]];
        if (!branch.size()) {
            bool null_root = spherical_root? p==mnpos: p==mnpos||p==0;
            branch.parent_id = null_root? mnpos: bids[p];                                                                                                                                             

            // Add the first sample to a branch that is attached to
            // non-spherical root if the branch is not branch 0.
            if (p!=mnpos && !(p==0 && spherical_root)) {
                branch.index.push_back(p);
            }   
        }   

        branch.index.push_back(i);
```

Reply:
Ah! I see; thanks for explaining. This is ultimately so we can avoid having to distinguish between being attached to the proximal or distal ends of a parent branch, right?

I'd only ask then for a unit test that checks this situation.

Reply:
There are unit tests that test that situation! That is how I know it works.

Reply:
But I tried the example code and the unit tests passed.

Reply:
It failed when I tried it!
`morphology.branches_from_parent_index` starts with tests on single sample trees, and progressively makes them more complex, testing both with and without `spherical_root`. The `morphology.branches` tests also failed in some cases.

Reply:
Okay, I am super confused. Maybe I was trying against the wrong branch? Will have a check when next at computer.

Reply:
I have no idea what I was testing before, because it is going exactly as you say.